pkg/network/prober/prober.go

```diff
@@ -20,31 +20,96 @@ import (
 	"context"
 	"io/ioutil"
 	"net/http"
+	"sync"
+	"time"
 
 	"github.com/pkg/errors"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/apimachinery/pkg/util/wait"
 
 	"github.com/knative/serving/pkg/network"
 )
 
 // Do sends a single probe to given target, e.g. `http://revision.default.svc.cluster.local:81`.
 // headerValue is the value for the `k-network-probe` header.
-// Do returns the status code, response body, and the request error, if any.
-func Do(ctx context.Context, target, headerValue string) (int, string, error) {
+// Do returns whether the probe was successful or not, or there was an error probing.
+func Do(ctx context.Context, target, headerValue string) (bool, error) {
 	req, err := http.NewRequest(http.MethodGet, target, nil)
 	if err != nil {
-		return 0, "", errors.Wrapf(err, "%s is not a valid URL", target)
+		return false, errors.Wrapf(err, "%s is not a valid URL", target)
 	}
 
 	req.Header.Set(http.CanonicalHeaderKey(network.ProbeHeaderName), headerValue)
 	req = req.WithContext(ctx)
 	resp, err := network.AutoTransport.RoundTrip(req)
 	if err != nil {
-		return 0, "", err
+		return false, errors.Wrapf(err, "error roundtripping %s", target)
 	}
 	defer resp.Body.Close()
 	body, err := ioutil.ReadAll(resp.Body)
 	if err != nil {
-		return 0, "", err
+		return false, errors.Wrap(err, "error reading body")
 	}
-	return resp.StatusCode, string(body), nil
+	return resp.StatusCode == http.StatusOK && string(body) == headerValue, nil
+}
+
+// Done is a callback that is executed when the async probe has finished.
+// `arg` is given by the caller at the offering time, while `success` and `err`
+// are the return values of the `Do` call.
+type Done func(arg interface{}, success bool, err error)
+
+// Manager manages async probes and makes sure we run concurrently only a single
+// probe for the same key.
+type Manager struct {
+	cb Done
+
+	// mu guards keys.
+	mu   sync.Mutex
+	keys sets.String
+}
+
+// New creates a new Manager, that will invoke the given callback when
+// async probing is finished.
+func New(cb Done) *Manager {
+	return &Manager{
+		keys: sets.NewString(),
+		cb:   cb,
+	}
+}
+
+// Offer executes asynchronous probe using `target` as the key.
+// If a probe with the same key already exists, Offer will return false and the
+// call is discarded. If the request is accepted, Offer returns true.
+// Otherwise Offer starts a goroutine that periodically executes
+// `Do`, until timeout is reached, the probe succeeds, or fails with an error.
+// In the end the callback is invoked with the provided `arg` and probing results.
+func (m *Manager) Offer(ctx context.Context, target, headerValue string, arg interface{}, period, timeout time.Duration) bool {
```

Comment:
Why "Offer"?

Reply:
we offer system to make a probe, since the system can reject doing it  -- it's not `Do` :)